eras/conway/impl/src/Cardano/Ledger/Conway/State/CertState.hs

```diff
@@ -0,0 +1,275 @@
+vsActualDRepExpiry cred vs =
+  binOpEpochNo (+) (vsNumDormantEpochs vs) . drepExpiry <$> Map.lookup cred (vsDReps vs)
+
+csCommitteeCredsL ::
+  Lens' (CommitteeState era) (Map (Credential 'ColdCommitteeRole) CommitteeAuthorization)
+csCommitteeCredsL = lens csCommitteeCreds (\cs u -> cs {csCommitteeCreds = u})
+
+epochStateRegDrepL ::
+  ConwayEraCertState era => Lens' (EpochState era) (Map (Credential 'DRepRole) DRepState)
+epochStateRegDrepL = esLStateL . lsCertStateL . certVStateL . vsDRepsL
+
+class EraCertState era => ConwayEraCertState era where
+  certVStateL :: Lens' (CertState era) (VState era)
+
+mkConwayCertState :: VState era -> PState era -> DState era -> ConwayCertState era
+mkConwayCertState v p d =
+  ConwayCertState
+    { conwayCertVState = v
+    , conwayCertPState = p
+    , conwayCertDState = d
+    }
+
+conwayCertDStateL :: Lens' (ConwayCertState era) (DState era)
+conwayCertDStateL = lens conwayCertDState (\ds u -> ds {conwayCertDState = u})
+{-# INLINE conwayCertDStateL #-}
+
+conwayCertPStateL :: Lens' (ConwayCertState era) (PState era)
+conwayCertPStateL = lens conwayCertPState (\ds u -> ds {conwayCertPState = u})
+{-# INLINE conwayCertPStateL #-}
+
+conwayCertVStateL :: Lens' (ConwayCertState era) (VState era)
+conwayCertVStateL = lens conwayCertVState (\ds u -> ds {conwayCertVState = u})
+{-# INLINE conwayCertVStateL #-}
+
+toCertStatePairs :: KeyValue e a => ConwayCertState era -> [a]
+toCertStatePairs certState@(ConwayCertState _ _ _) =
+  let ConwayCertState {..} = certState
+   in [ "dstate" .= conwayCertDState
+      , "pstate" .= conwayCertPState
+      , "vstate" .= conwayCertVState
+      ]
+
+conwayObligationCertState :: ConwayCertState era -> Obligations
+conwayObligationCertState (ConwayCertState VState {vsDReps} PState {psDeposits} DState {dsUnified}) =
+  let accum ans drepState = ans <> drepDeposit drepState
+   in Obligations
+        { oblStake = UM.fromCompact (UM.sumDepositUView (UM.RewDepUView dsUnified))
+        , oblPool = F.foldl' (<>) (Coin 0) psDeposits
+        , oblDRep = F.foldl' accum (Coin 0) vsDReps
+        , oblProposal = Coin 0
+        }
+
+conwayCertsTotalDepositsTxBody ::
+  EraTxBody era => PParams era -> ConwayCertState era -> TxBody era -> Coin
+conwayCertsTotalDepositsTxBody pp ConwayCertState {conwayCertPState} =
+  getTotalDepositsTxBody pp (`Map.member` psStakePoolParams conwayCertPState)
+
+conwayCertsTotalRefundsTxBody ::
+  EraTxBody era => PParams era -> ConwayCertState era -> TxBody era -> Coin
+conwayCertsTotalRefundsTxBody pp ConwayCertState {conwayCertDState, conwayCertVState} =
+  getTotalRefundsTxBody
+    pp
+    (lookupDepositDState conwayCertDState)
+    (lookupDepositVState conwayCertVState)
+
+instance EraCertState ConwayEra where
+  type CertState ConwayEra = ConwayCertState ConwayEra
+
+  -- TODO: use the Translation interface instead
+  upgradeCertState (ShelleyCertState p d) = ConwayCertState def (coerce p) (coerce d)
```

Comment:
I think you can safely remove `upgradeCertState` function, cause I can't see any place where we could use it.